mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -5390,85 +5672,152 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
     // llvm::errs() << "KPerBlock: " << KPerBlock << "\n";
     // llvm::errs() << "MPerBlock: " << MPerBlock << "\n";
     // llvm::errs() << "NPerBlock: " << NPerBlock << "\n";
-    // llvm::errs() << "matrix_a_source_data_per_read: " << matrix_a_source_data_per_read << "\n";
-    // llvm::errs() << "matrix_b_source_data_per_read: " << matrix_b_source_data_per_read << "\n";
+    // llvm::errs() << "KPack: " << KPack << "\n";
+    // llvm::errs() << "matrix_a_source_data_per_read: "
+    //              << matrix_a_source_data_per_read << "\n";
+    // llvm::errs() << "matrix_b_source_data_per_read: "
+    //              << matrix_b_source_data_per_read << "\n";
 
     // Compute ThreadSliceLengths for Matrix A.
     uint64_t GemmABlockCopyNumberDataPerThread =
         MPerBlock * KPerBlock / BlockSize;
 
     uint64_t GemmABlockCopyThreadSliceLengths_GemmK;
     uint64_t GemmABlockCopyThreadSliceLengths_GemmM;
+    uint64_t GemmABlockCopyThreadSliceLengths_GemmKPack = 1;
     switch (matrix_a_source_vector_read_dim) {
     case GemmK:
-      GemmABlockCopyThreadSliceLengths_GemmK = matrix_a_source_data_per_read;
-      GemmABlockCopyThreadSliceLengths_GemmM =
-          GemmABlockCopyNumberDataPerThread /
-          GemmABlockCopyThreadSliceLengths_GemmK;
+      if (KPack > 1) {
+        GemmABlockCopyThreadSliceLengths_GemmKPack =
+            matrix_a_source_data_per_read;
+        GemmABlockCopyThreadSliceLengths_GemmK =
+            KPack / matrix_a_source_data_per_read;
+        GemmABlockCopyThreadSliceLengths_GemmM =
+            GemmABlockCopyNumberDataPerThread / KPack;
+      } else {
+        GemmABlockCopyThreadSliceLengths_GemmK = matrix_a_source_data_per_read;
+        GemmABlockCopyThreadSliceLengths_GemmM =
+            GemmABlockCopyNumberDataPerThread /
+            GemmABlockCopyThreadSliceLengths_GemmK;
+      }
       break;
     case GemmMorN:
-      GemmABlockCopyThreadSliceLengths_GemmM = matrix_a_source_data_per_read;
-      GemmABlockCopyThreadSliceLengths_GemmK =
-          GemmABlockCopyNumberDataPerThread /
-          GemmABlockCopyThreadSliceLengths_GemmM;
+      // TBD: FIXME. Review logic here.
+      if (KPack > 1) {
```

Comment:
Something doesn't feel right about the logic here. Can you talk me through what these variables do, how they're computed, and how they're affected by kpack?